bpf/xdp/attach.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2020-2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xdp
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/projectcalico/felix/bpf"
+	log "github.com/sirupsen/logrus"
+)
+
+type AttachPoint struct {
+	Iface    string
+	LogLevel string
+	Modes    []bpf.XDPMode
+}
+
+func (ap *AttachPoint) IfaceName() string {
+	return ap.Iface
+}
+
+func (ap *AttachPoint) JumpMapFDMapKey() string {
+	return "xdp"
+}
+
+func (ap *AttachPoint) FileName() string {
+	logLevel := strings.ToLower(ap.LogLevel)
+	if logLevel == "off" {
+		logLevel = "no_log"
+	}
+	return "xdp_" + logLevel + ".o"
+}
+
+func (ap *AttachPoint) SectionName() string {
+	return "calico_entrypoint_xdp"
+}
+
+func (ap *AttachPoint) Log() *log.Entry {
+	return log.WithFields(log.Fields{
+		"iface":    ap.Iface,
+		"modes":    ap.Modes,
+		"logLevel": ap.LogLevel,
+	})
+}
+
+func (ap *AttachPoint) AttachProgram() error {
+	preCompiledBinary := path.Join(bpf.ObjectDir, ap.FileName())
+	sectionName := ap.SectionName()
+
+	// Patch the binary so that its log prefix is like "eth0------X".
+	tempDir, err := ioutil.TempDir("", "calico-tc")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary directory: %w", err)
+	}
+	defer func() {
+		_ = os.RemoveAll(tempDir)
+	}()
+	tempBinary := path.Join(tempDir, ap.FileName())
+	err = ap.patchBinary(preCompiledBinary, tempBinary)
+	if err != nil {
+		ap.Log().WithError(err).Error("Failed to patch binary")
+		return err
+	}
+
+	var errs []error
+	for _, mode := range ap.Modes {
+		ap.Log().Infof("Attempt XDP attach with mode %v", mode)
+
+		// First remove any existing program.
+		cmd := exec.Command("ip", "link", "set", "dev", ap.Iface, mode.String(), "off")
+		ap.Log().Infof("Running: %v %v", cmd.Path, cmd.Args)
+		out, err := cmd.CombinedOutput()
+		ap.Log().WithField("mode", mode).Infof("Result: err=%v out=\n%v", err, string(out))
```

Comment:
It is possible to attach XDP/BPF programs to different modes at the same time. For example, one program can be attached to a device both as native mode and generic mode at the same time. We should make sure all loaded programs are unloaded, event if an attempt to attach succeed.

Reply:
Demoted all logs to Debug except this one, which should be just once per host interface - WDYT?
```
			ap.Log().Infof("Successful XDP attachment with mode %v", mode)
```

Reply:
On `-force` and `off` I think there are a few considerations.

Firstly, if we use `-force`, and it fails because of a problem with the new program, I believe an old program (if any) will remain attached - which means we need to use `off` if `-force` fails.  In that case, I think it's equally nice to do `off` first and then `ip link set dev xdp ...`.

Secondly, dealing with other XDP programs that might be there at start of day.  (Either 3rd party, or maybe some earlier version of Felix.)  I think we fundamentally have to clean those up, as I don't believe our XDP usage can coexist with anyone else's.

Thirdly, for a given host interface, is it possible for our best attachment mode to change, other than at start of day?  I think that would only be possible if we did a patch that would alter some conditionally included code.  We don't have that for XDP right now, but we conceivably could in future.  So probably we should assume it's possible for the attachment mode to improve, which means needing to do `off` for subsequent modes.

So I think the result is:
- Loop through the modes.
- Always do `off`
- If we haven't yet successfully attached (for this loop), do the attachment.
- Continue through remaining modes even if attachment successful.

Reply:
My thinking is that, in the upgrade case (or felix restart case) we don't want to flap the program even for a short time.  The program could be protecting the host from a packet flood and it might take a few hundred ms to get the new program in place.  We should avoid that if we can.

Reply:
Ah right, I didn't think of that.  OK, will tweak to use `-force` then.